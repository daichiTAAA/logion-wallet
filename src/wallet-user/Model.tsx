
export interface TokenizationRequest {
    legalOfficerAddress: string,
    requesterAddress: string,
    requestedTokenName: string,
    bars: number,
}
