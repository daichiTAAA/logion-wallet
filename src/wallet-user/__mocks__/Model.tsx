import { TokenizationRequest } from "../../common/types/ModelTypes";
import { CreateTokenRequest } from "../Model";

export async function createTokenRequest(request: CreateTokenRequest): Promise<TokenizationRequest> {
    return Promise.resolve({
        id: "1234",
        legalOfficerAddress: request.legalOfficerAddress,
        requesterAddress: request.requesterAddress,
        requestedTokenName: request.requestedTokenName,
        bars: request.bars,
        status: "PENDING"
    });
}
