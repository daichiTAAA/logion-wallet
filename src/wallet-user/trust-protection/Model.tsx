import axios from "axios";
import {Moment} from "moment";
import {DEFAULT_LEGAL_OFFICER, ANOTHER_LEGAL_OFFICER, ProtectionRequest} from "../../legal-officer/Types";
import Identity from '../../component/types/Identity';
import PostalAddress from '../../component/types/PostalAddress';

export interface CreateProtectionRequest {
    requesterAddress: string,
    userIdentity: Identity,
    userPostalAddress: PostalAddress,
    legalOfficerAddresses: string[],
    signature: string,
    signedOn: Moment
}

export type LegalOfficerDecisionStatus = "PENDING" | "REJECTED" | "ACCEPTED";

export interface LegalOfficerDecision {
    legalOfficerAddress: string,
    status: LegalOfficerDecisionStatus
}

export async function createProtectionRequest(request: CreateProtectionRequest): Promise<ProtectionRequest> {
    const response = await axios.post("/api/protection-request", request);
    return response.data;
}

function fakeAddress(i: number) {
    return "1234567890-abcd-" + i;
}

export const legalOfficers = [
    {name: "Patrick", address: DEFAULT_LEGAL_OFFICER},
    {name: "Guillaume", address: ANOTHER_LEGAL_OFFICER},
    {name: "Alain", address: fakeAddress(5)},
    {name: "Ronald", address: fakeAddress(6)},
    {name: "Thibaut", address: fakeAddress(7)},
    {name: "Jona", address: fakeAddress(8)},
    {name: "Carlos", address: fakeAddress(9)},
    {name: "Ilias", address: fakeAddress(10)},
    {name: "Pedroso", address: fakeAddress(11)},
    {name: "Marc", address: fakeAddress(12)},
]

export function legalOfficerName(address: string): string {
    const legalOfficer = legalOfficers.find(legalOfficer => legalOfficer.address === address);
    return legalOfficer === undefined ? "Unknown" : legalOfficer.name;
}

