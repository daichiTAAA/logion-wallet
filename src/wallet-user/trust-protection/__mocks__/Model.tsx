export function checkActivation() {
    return Promise.resolve();
}
