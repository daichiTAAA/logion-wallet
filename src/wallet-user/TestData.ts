import { Coin, CoinBalance } from '../logion-chain/Balances';
import { PrefixedNumber, ATTO } from '../logion-chain/numbers';
import { Transaction } from './Types';

export const TEST_WALLET_USER = "5H4MvAsobfZ6bBCDyj5dsrWYLrA8HrRzaqa9p61UXtxMhSCY";

export const DEFAULT_COIN: Coin = {
    id: 'log',
    iconId: 'log',
    iconType: 'png',
    name: "Logion",
    symbol: "LOG"
};

export const DEFAULT_COIN_BALANCE: CoinBalance = {
    coin: DEFAULT_COIN,
    balance: new PrefixedNumber("42", ATTO),
    level: 0.1,
};

export const DEFAULT_TRANSACTION: Transaction = {
    from: "from",
    to: "to",
    pallet: "pallet",
    method: "method",
    transferValue: "420",
    tip: "0",
    fee: "125000",
    reserved: "0",
    total: "125420",
    createdOn: "2021-07-28T12:30:00.000",
    type: "Sent",
}
