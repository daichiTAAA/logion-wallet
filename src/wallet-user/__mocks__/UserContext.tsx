import {DEFAULT_LEGAL_OFFICER, TokenizationRequest} from "../../legal-officer/Types";
import {TEST_WALLET_USER} from "../Model.test";
import {ApiPromise} from '@polkadot/api';
import {ProtectionRequest} from "../../legal-officer/Types";
import { DARK_MODE } from '../Types';

export let createTokenRequest = () => null;

export let createdTokenRequest: TokenizationRequest | null = null;

export let pendingTokenizationRequests: TokenizationRequest[] | null = null;

export let acceptedTokenizationRequests: TokenizationRequest[] | null = null;

export let rejectedTokenizationRequests: TokenizationRequest[] | null = null;

export let refreshRequests = jest.fn();

export let createProtectionRequest = () => null;

export let pendingProtectionRequests: ProtectionRequest[] | null = null;

export let acceptedProtectionRequests: ProtectionRequest[] | null = null;

export let api = new ApiPromise();

export let recoveryConfig = null;

export let setUserAddress = jest.fn();

export let userAddress = TEST_WALLET_USER;

export function useUserContext() {
    return {
        legalOfficerAddress: DEFAULT_LEGAL_OFFICER,
        userAddress,
        createTokenRequest,
        createdTokenRequest,
        pendingTokenizationRequests,
        acceptedTokenizationRequests,
        rejectedTokenizationRequests,
        refreshRequests,
        createProtectionRequest,
        pendingProtectionRequests,
        acceptedProtectionRequests,
        recoveryConfig,
        api,
        setUserAddress,
        colorTheme: DARK_MODE,
    };
}

export function setCreateTokenRequest(callback: any) {
    createTokenRequest = callback;
}

export function setCreatedTokenRequest(request: TokenizationRequest) {
    createdTokenRequest = request;
}

export function setPendingRequests(requests: TokenizationRequest[]) {
    pendingTokenizationRequests = requests;
}

export function setRejectedRequests(requests: TokenizationRequest[]) {
    rejectedTokenizationRequests = requests;
}

export function setAcceptedRequests(requests: TokenizationRequest[]) {
    acceptedTokenizationRequests = requests;
}

export function setRefreshRequests(func: any) {
    refreshRequests = func;
}

export function setPendingProtectionRequests(requests: ProtectionRequest[] | null) {
    pendingProtectionRequests = requests;
}

export function setAcceptedProtectionRequests(requests: ProtectionRequest[] | null) {
    acceptedProtectionRequests = requests;
}

export function setCreateProtectionRequest(callback: any) {
    createProtectionRequest = callback;
}

export function setRecoveryConfig(config: any) {
    recoveryConfig = config;
}

export function UserContextProvider() {
    return (
        <div></div>
    );
}

export function setContextUserAddress(address: string) {
    userAddress = address;
}