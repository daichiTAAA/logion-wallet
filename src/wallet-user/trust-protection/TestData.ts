import moment from 'moment';

import { ProtectionRequest, ProtectionRequestStatus } from '../../common/types/ModelTypes';
import { DEFAULT_IDENTITY, DEFAULT_ADDRESS } from '../../common/TestData';
import { TEST_WALLET_USER } from '../TestData';

export const ACTIVATED_PROTECTION_REQUEST: ProtectionRequest = createProtectionRequest("ACTIVATED", false, null);

export const PENDING_PROTECTION_REQUEST: ProtectionRequest = createProtectionRequest("PENDING", false, null);

export const PROTECTION_REQUEST: ProtectionRequest = ACTIVATED_PROTECTION_REQUEST;

export const ACTIVATED_RECOVERY_REQUEST: ProtectionRequest = createProtectionRequest("ACTIVATED", true, "toRecover");

export const PENDING_RECOVERY_REQUEST: ProtectionRequest = createProtectionRequest("PENDING", true, "toRecover");

export const RECOVERY_REQUEST: ProtectionRequest = ACTIVATED_RECOVERY_REQUEST;

function createProtectionRequest(status: ProtectionRequestStatus, isRecovery:boolean, addressToRecover:string | null): ProtectionRequest {
    return {
        id: '2eb5f71c-7f31-44b5-9390-c3bf56501880',
        requesterAddress: TEST_WALLET_USER,
        decisions: [
            {
                legalOfficerAddress: '5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty',
                status: 'PENDING',
                rejectReason: null,
                decisionOn: null,
                requestId: '2eb5f71c-7f31-44b5-9390-c3bf56501880'
            },
            {
                legalOfficerAddress: '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY',
                status: 'PENDING',
                rejectReason: null,
                decisionOn: null,
                requestId: '2eb5f71c-7f31-44b5-9390-c3bf56501880'
            }
        ],
        userIdentity: DEFAULT_IDENTITY,
        userPostalAddress: DEFAULT_ADDRESS,
        createdOn: moment('2021-06-10T13:48:00.000Z').toISOString(),
        isRecovery,
        addressToRecover,
        status,
    };
}
