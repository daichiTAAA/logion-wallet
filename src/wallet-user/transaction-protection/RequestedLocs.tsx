import React, { useCallback, useEffect, useState } from 'react';
import { useForm } from 'react-hook-form';
import { Option } from '@polkadot/types';

import { useCommonContext } from '../../common/CommonContext';
import Table, { Cell, EmptyTableMessage, DateCell } from '../../common/Table';
import LocStatusCell from '../../common/LocStatusCell';
import Button from '../../common/Button';
import Dialog from '../../common/Dialog';
import { CreateLocRequest, createLocRequest } from '../../common/Model';
import UserIdentity from '../../common/types/Identity';
import { useUserContext } from '../UserContext';
import { RecoveryConfig } from '../../logion-chain/Recovery';

import LocCreationForm, { FormValues } from './LocCreationForm';

function shouldShowIdentityFields(
    legalOfficer: string | null,
    recoveryConfig: Option<RecoveryConfig> | null
): boolean {
    if(legalOfficer === null || recoveryConfig === null || recoveryConfig.isNone) {
        return true;
    } else {
        return !recoveryConfig.unwrap().friends.toArray().map(accountId => accountId.toString()).includes(legalOfficer);
    }
}

export default function RequestedLocs() {
    const { colorTheme, pendingLocRequests, accounts, refresh, axios } = useCommonContext();
    const [ requestLoc, setRequestLoc ] = useState(false);
    const { control, handleSubmit, formState: { errors }, reset, watch } = useForm<FormValues>({
        defaultValues: {
            description: "",
            legalOfficer: "",
            firstName: "",
            lastName: "",
            email: "",
            phone: ""
        }
    });
    const [ selectedLegalOfficer, setSelectedLegalOfficer ] = useState<string | null>(null);
    const { recoveryConfig } = useUserContext();

    const showIdentityFields = shouldShowIdentityFields(selectedLegalOfficer, recoveryConfig);

    const submit = useCallback((formValues: FormValues) => {
        let userIdentity: UserIdentity | undefined;
        if(showIdentityFields) {
            userIdentity = {
                firstName: formValues.firstName || "",
                lastName: formValues.lastName || "",
                email: formValues.email || "",
                phoneNumber: formValues.phone || "",
            };
        } else {
            userIdentity = undefined;
        }

        (async function() {
            const currentAddress = accounts!.current!.address;
            const request: CreateLocRequest = {
                ownerAddress: formValues.legalOfficer,
                requesterAddress: currentAddress,
                description: formValues.description,
                userIdentity,
            }
            await createLocRequest!(axios!, request);
            reset();
            setRequestLoc(false);
            refresh!();
        })();
    }, [ axios, accounts, setRequestLoc, refresh, reset, showIdentityFields ]);

    useEffect(() => {
        const subscription = watch(({ legalOfficer }) => setSelectedLegalOfficer(legalOfficer));
        return () => subscription.unsubscribe();
    }, [watch]);

    if(pendingLocRequests === null) {
        return null;
    }

    return (
        <>
            <Table
                columns={[
                    {
                        header: "Legal officer",
                        render: request => <Cell content={ request.ownerAddress }  overflowing tooltipId={ `dest-${request.id}` } />,
                        align: 'left',
                    },
                    {
                        "header": "Description",
                        render: request => <Cell content={ request.description } />,
                        align: 'left',
                    },
                    {
                        header: "Status",
                        render: request => <LocStatusCell status={ request.status }/>,
                        width: "140px",
                    },
                    {
                        header: "Creation date",
                        render: request => <DateCell dateTime={ request.createdOn || null } />,
                        width: '200px',
                    }
                ]}
                data={ pendingLocRequests }
                renderEmpty={ () => <EmptyTableMessage>No requested LOCs</EmptyTableMessage> }
            />
            <Button onClick={() => setRequestLoc(true)}>Request new LOC</Button>
            <Dialog
                show={ requestLoc }
                size="lg"
                actions={[
                    {
                        id: "submit",
                        buttonText: 'Submit',
                        buttonVariant: 'primary',
                        type: 'submit',
                    },
                    {
                        id: "cancel",
                        callback: () => { reset() ; setRequestLoc(false) },
                        buttonText: 'Cancel',
                        buttonVariant: 'secondary',
                    }
                ]}
                onSubmit={handleSubmit(submit)}
            >
                <LocCreationForm
                    control={ control }
                    errors={ errors }
                    colors={ colorTheme.dialog }
                    legalOfficer={ selectedLegalOfficer }
                    showIdentityFields={ showIdentityFields }
                />
            </Dialog>
        </>
    );
}
