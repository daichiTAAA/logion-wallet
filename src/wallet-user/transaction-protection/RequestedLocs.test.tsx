jest.mock('../../common/CommonContext');
jest.mock('../../common/Model');

import {fireEvent, render, screen, waitFor} from '@testing-library/react';

import { shallowRender } from '../../tests';
import RequestedLocs from './RequestedLocs';
import { axiosMock, setPendingLocRequests } from '../../common/__mocks__/CommonContextMock';
import { createLocRequest } from '../../common/Model';
import { DEFAULT_LEGAL_OFFICER } from '../../common/types/LegalOfficer';
import { TEST_WALLET_USER } from '../TestData';

it("Renders null with no data", () => {
    const tree = shallowRender(<RequestedLocs />);
    expect(tree).toMatchSnapshot();
});

it("Renders pending requests", () => {
    setPendingLocRequests([
        {
            id: "1",
            ownerAddress: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
            requesterAddress: "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
            description: "LOC description",
            status: "REQUESTED"
        }
    ]);
    const tree = shallowRender(<RequestedLocs/>);
    expect(tree).toMatchSnapshot();
});

it("should display messages when an empty form is submitted", async () => {
    render(<RequestedLocs/>);
    openDialog();

    const button = screen.getByRole('button', {name: "Submit"});
    fireEvent.click(button)

    await waitFor(() => expect(screen.getByText("The description is required")).toBeInTheDocument());
});

function openDialog() {
    const button = screen.getByRole('button', {name: "Request new LOC"});
    fireEvent.click(button);
}

it("should create a LOC and display no message when a valid form is submitted", async () => {
    render(<RequestedLocs/>);
    openDialog();

    const description = "description";
    const descriptionInput = screen.getByRole("textbox", {name: "Description"});
    fireEvent.input(descriptionInput, {target: {value: description}})

    const button = screen.getByRole('button', {name: "Submit"});
    fireEvent.click(button);

    await waitFor(() => expect(createLocRequest).toBeCalledWith(
        axiosMock.object(),
        expect.objectContaining({
            ownerAddress: DEFAULT_LEGAL_OFFICER,
            requesterAddress: TEST_WALLET_USER,
            description,
        })
    ));
});

it("should close dialog and not create token when cancel is pressed", async () => {
    render(<RequestedLocs/>);
    openDialog();

    const dialog = screen.getByRole("dialog");
    const button = screen.getByRole('button', {name: "Cancel"});
    fireEvent.click(button)
    expect(createLocRequest).not.toBeCalled();
    await waitFor(() => expect(dialog).not.toBeInTheDocument());
});
