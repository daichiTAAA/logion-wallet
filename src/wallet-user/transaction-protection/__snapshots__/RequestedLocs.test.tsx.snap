// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders null with no data 1`] = `null`;

exports[`Renders pending requests 1`] = `
<React.Fragment>
  <Table
    columns={
      Array [
        Object {
          "align": "left",
          "header": "Legal officer",
          "render": [Function],
        },
        Object {
          "align": "left",
          "header": "Description",
          "render": [Function],
        },
        Object {
          "header": "Status",
          "render": [Function],
          "width": "140px",
        },
        Object {
          "header": "Creation date",
          "render": [Function],
          "width": "200px",
        },
      ]
    }
    data={
      Array [
        Object {
          "description": "LOC description",
          "id": "1",
          "ownerAddress": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "requesterAddress": "5Ew3MyB15VprZrjQVkpQFj8okmc9xLDSEdNhqMMS5cXsqxoW",
          "status": "REQUESTED",
        },
      ]
    }
    renderEmpty={[Function]}
  />
  <Button
    onClick={[Function]}
  >
    Request new LOC
  </Button>
  <Dialog
    actions={
      Array [
        Object {
          "buttonText": "Submit",
          "buttonVariant": "primary",
          "id": "submit",
          "type": "submit",
        },
        Object {
          "buttonText": "Cancel",
          "buttonVariant": "secondary",
          "callback": [Function],
          "id": "cancel",
        },
      ]
    }
    onSubmit={[Function]}
    show={false}
    size="lg"
  >
    <LocCreationForm
      colors={
        Object {
          "background": "#eff3fe",
          "borderColor": "#e6007a",
          "foreground": "#000000",
        }
      }
      control={
        Object {
          "_defaultValues": Object {
            "description": "",
            "email": "",
            "firstName": "",
            "lastName": "",
            "legalOfficer": "",
            "phone": "",
          },
          "_fields": Object {},
          "_formState": Object {
            "dirtyFields": Object {},
            "errors": Object {},
            "isDirty": false,
            "isSubmitSuccessful": false,
            "isSubmitted": false,
            "isSubmitting": false,
            "isValid": false,
            "isValidating": false,
            "submitCount": 0,
            "touchedFields": Object {},
          },
          "_formValues": Object {},
          "_getFieldArrayValue": [Function],
          "_getIsDirty": [Function],
          "_getWatch": [Function],
          "_isInAction": false,
          "_isMounted": false,
          "_names": Object {
            "array": Set {},
            "mount": Set {},
            "unMount": Set {},
            "watch": Set {},
          },
          "_proxyFormState": Object {
            "dirtyFields": false,
            "errors": "all",
            "isDirty": false,
            "isValid": false,
            "isValidating": false,
            "touchedFields": false,
          },
          "_removeFields": [Function],
          "_shouldUnregister": undefined,
          "_subjects": Object {
            "array": ge {
              "observers": Array [],
            },
            "control": ge {
              "observers": Array [],
            },
            "state": ge {
              "observers": Array [],
            },
            "watch": ge {
              "observers": Array [],
            },
          },
          "_updateFieldArray": [Function],
          "_updateProps": [Function],
          "_updateValid": [Function],
          "_updateValues": [Function],
          "register": [Function],
          "unregister": [Function],
        }
      }
      errors={Object {}}
      legalOfficer={null}
      showIdentityFields={true}
    />
  </Dialog>
</React.Fragment>
`;
